src/libsystemd/sd-journal/journal-file.c

```diff
@@ -1927,10 +1924,16 @@ static int journal_file_append_entry_internal(
         if (boot_id)
                 f->header->boot_id = *boot_id;
 
+        if (f->header->head_entry_realtime == 0)
+                f->header->head_entry_realtime = htole64(ts->realtime);
+
         if (JOURNAL_HEADER_COMPACT(f->header)) {
+                if (f->header->head_entry_monotonic == 0)
+                        f->header->head_entry_monotonic = htole64(ts->monotonic);
+
                 o->entry.compact.seqnum = htole64(journal_file_new_entry_seqnum(f, seqnum));
-                o->entry.compact.realtime = htole64(ts->realtime);
-                o->entry.compact.monotonic = htole64(ts->monotonic);
+                o->entry.compact.realtime = htole32(ts->realtime - le64toh(f->header->head_entry_realtime));
+                o->entry.compact.monotonic = htole32(ts->monotonic - le64toh(f->header->head_entry_monotonic));
```

Comment:
Shouldn't this raise an error so that the file is rotated? Or at least saturate to the maximum value to avoid wraparound? Let's say that we're appending to a file which was created on a system with no RTC, so the first few entries are at time close to zero, and then we start getting entries with normal timestamps…

Reply:
Added some logic to check that the timestamps aren't too early or too far in the future. We return an error if that's the case